lib/teiserver/o_auth.ex

```diff
@@ -0,0 +1,386 @@
+  The token scopes are the same as the application
+  """
+  @spec create_code(
+          User.t() | T.userid(),
+          %{
+            id: integer(),
+            scopes: Application.scopes(),
+            redirect_uri: String.t(),
+            challenge: String.t(),
+            challenge_method: String.t()
+          },
+          options()
+        ) ::
+          {:ok, Code.t()} | {:error, Ecto.Changeset.t()}
+  def create_code(user, attrs, opts \\ [])
+
+  def create_code(%User{} = user, attrs, opts) do
+    create_code(user.id, attrs, opts)
+  end
+
+  def create_code(user, attrs, opts) when is_map(user) do
+    create_code(user.id, attrs, opts)
+  end
+
+  def create_code(user_id, attrs, opts) do
+    now = Keyword.get(opts, :now, Timex.now())
+
+    # don't do any validation on the challenge yet, this is done when exchanging
+    # the code for a token
+    attrs = %{
+      value: Base.hex_encode32(:crypto.strong_rand_bytes(32)),
+      owner_id: user_id,
+      application_id: attrs.id,
+      scopes: attrs.scopes,
+      expires_at: Timex.add(now, Timex.Duration.from_minutes(5)),
+      redirect_uri: Map.get(attrs, :redirect_uri),
+      challenge: Map.get(attrs, :challenge),
+      challenge_method: Map.get(attrs, :challenge_method)
+    }
+
+    %Code{}
+    |> Code.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Given a code returns the corresponding db object, making sure
+  it is valid (exists, not expired, not revoked)
+  """
+  @spec get_valid_code(String.t(), options()) :: {:ok, Code.t()} | {:error, term()}
+  def get_valid_code(code, opts \\ [])
+
+  def get_valid_code(code, opts) do
+    case CodeQueries.get_code(code) do
+      nil ->
+        {:error, :no_code}
+
+      code ->
+        now = Keyword.get(opts, :now, Timex.now())
+
+        if expired?(code, now) do
+          {:error, :expired}
+        else
+          {:ok, code}
+        end
+    end
+  end
+
+  @spec create_token(
+          User.t() | T.userid(),
+          %{id: integer(), scopes: Application.scopes()},
+          options()
+        ) ::
+          {:ok, Token.t()} | {:error, Ecto.Changeset.t()}
+  def create_token(user_id, application, opts \\ [])
+
+  def create_token(%User{} = user, application, opts) do
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user, application, opts) when is_map(user) do
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user_id, application, opts) do
+    do_create_token(%{owner_id: user_id}, application, opts)
+  end
+
+  defp do_create_token(owner_attr, application, opts \\ []) do
+    now = Keyword.get(opts, :now, DateTime.utc_now())
+
+    refresh_attrs =
+      %{
+        value: Base.hex_encode32(:crypto.strong_rand_bytes(32), padding: false),
+        application_id: application.id,
+        scopes: application.scopes,
+        expires_at: Timex.add(now, Timex.Duration.from_days(60)),
```

Comment:
I think this should be *much* longer then 60, IMHO it should be almost forever: https://discord.com/channels/549281623154229250/724293659809284137/1162856143719243888

Reply:
Amended and bumped it to 100 years.